File: ide/src/main/java/org/dau/ide/main/MainProjects.java
package org.dau.ide.main;

import javafx.collections.ObservableSet;
import javafx.collections.SetChangeListener;
import javafx.collections.SetChangeListener.Change;
import org.dau.ui.schematic.fx.model.FxProject;
import org.dau.ui.schematic.fx.model.FxSchema;
import org.springframework.stereotype.Component;

import java.nio.file.Files;
import java.util.HashMap;

import static javafx.collections.FXCollections.observableSet;

@Component
public class MainProjects {

  private final MainDirectories directories;
  public final ObservableSet<FxProject> projects = observableSet();
  private final HashMap<FxProject, SetChangeListener<FxSchema>> schemaListeners = new HashMap<>();

  public MainProjects(MainDirectories directories) {
    this.directories = directories;
    this.projects.addListener(this::onChangeProjects);
  }

  private void onChangeProjects(Change<? extends FxProject> change) {
    if (change.wasAdded()) {
      var project = change.getElementAdded();
      final SetChangeListener<FxSchema> l = e -> onChangeSchema(project, e);
      project.schemas.addListener(l);
      schemaListeners.put(project, l);
    }
    if (change.wasRemoved()) {
      var project = change.getElementRemoved();
      var listener = schemaListeners.remove(project);
      project.schemas.removeListener(listener);
    }
  }

  private void onChangeSchema(FxProject project, Change<? extends FxSchema> change) {
    final var dir = directories.homeDir.resolve(project.id);
    if (change.wasRemoved()) {
      var schema = change.getElementRemoved();
      var file = dir.resolve(schema.toFileName());
      if (Files.isRegularFile(file)) {
        if (!file.toFile().delete()) {
          throw new IllegalStateException("Could not delete " + file);
        }
      }
    }
  }
}

File: ide/src/main/kotlin/org/dau/ide/main/MainProjects.kt
package org.dau.ide.main

import javafx.collections.FXCollections.observableSet
import javafx.collections.SetChangeListener
import javafx.collections.SetChangeListener.Change
import org.dau.ui.schematic.model.FxProject
import org.dau.ui.schematic.model.FxSchema
import org.springframework.stereotype.Component
import java.nio.file.Files

@Component
class MainProjects(private val directories: MainDirectories) {

  private val schemaListeners = HashMap<FxProject, SetChangeListener<FxSchema>>()
  val projects = observableSet<FxProject>().apply { addListener(SetChangeListener { onChangeProjects(it) }) }

  private fun onChangeProjects(change: Change<out FxProject>) {
    if (change.wasAdded()) {
      val project = change.elementAdded
      val l = SetChangeListener<FxSchema> { onChangeSchema(project, it) }
      project.schemas.addListener(l)
      schemaListeners[project] = l
    }
    if (change.wasRemoved()) {
      val project = change.elementRemoved
      val listener = schemaListeners.remove(project)
      project.schemas.removeListener(listener)
    }
  }

  private fun onChangeSchema(project: FxProject, change: Change<out FxSchema>) {
    val dir = directories.homeDir.resolve(project.id)
    if (change.wasRemoved()) {
      val schema = change.elementRemoved
      val file = dir.resolve(schema.toFileName())
      if (Files.isRegularFile(file)) {
        check(file.toFile().delete()) { "Could not delete $file" }
      }
    }
  }
}
